pico.py

```diff
@@ -6,43 +6,127 @@
+
+	if not is_valid_command(command):
+		return 0
+
+	return command >> len(bin(command)[2:]) - 4
+
+def get_command_data(command):
+	if not is_valid_command(command):
+		return 0
+	
+	header_and_data = command >> len(bin(command)[2:]) - 24
+	return header_and_data & int('0' * 4 + '1' * 20, 2)
+
 def signal_handler(signal, frame):
 	global connection
-	with connection:
-		print("\nSending closing signal...")
-		connection.sendall(b'\x00\x00\x00\x00')
-		sys.exit(0)
+	try:
+		with connection:
+			print("\nSending closing signal...")
+			connection.sendall(b'\x00\x00\x00\x00')
+	except:
+		print("\nNo client was connected")
+	sys.exit(0)
 
 signal.signal(signal.SIGINT, signal_handler)
 
 def main():
 	global connection
 
+	# This is just here to show how the above functions work
+	# It does nothing
+	command_example = 0b1000110011001100110011000011111
+	print(f"Header: {bin(get_command_header(command_example))}")
+	print(f"Data: {bin(get_command_data(command_example))}")
+
+	# Invalid example
+	get_command_header(0b11)
+
 	with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
 		s.bind((HOST, PORT))
 		s.listen()
 
 		print(f"Server is listening {s}")
 
-		print("Waiting for connection...")
+		while True:
+			print("Waiting for connection...")
 
-		connection, addr = s.accept()
+			connection, addr = s.accept()
 
-		with connection:
-			print(f"Connected {addr}")
-			while True:
-				data = connection.recv(4)
-				
-				if data == b'\x00\x00\x00\x00':
-					print("Closing signal received. Closing...")
-					break
-
-				binary_int = int.from_bytes(data, byteorder='big')
-				print(f"Received data: {bin(binary_int)}")
-				time.sleep(2)
-
-				print(f"Echoing data: {bin(binary_int)}")
-				connection.sendall(data)
+			with connection:
+				print(f"Connected {addr}")
+				while True:
+					try:
+						if is_connection_closed(connection):
+							print('Client disconnected')
+							connection.close()
+							break
+
+						data = connection.recv(4)
+						
+						if data == b'\x00\x00\x00\x00':
+							print("Closing signal received. Closing...")
+							connection.close()
+							break
+
+						binary_int = int.from_bytes(data, byteorder='big')
+
+						print(f"Received data: {bin(binary_int)}")
+
+						# Commands (fake examples obviously)
+						# Emit: 1111 / 15
+						# - Next 20 bits are data to emit
+						# Request Pico Status: 1010 / 10
+						# - Next 20 bits are ignored
+						# Enter power saving mode: 1001 / 9
+						# - Next 20 bits are ignored
+						command_header = get_command_header(binary_int)
+
+						if command_header == 15:
+							print(f"Emitting  data {bin(get_command_data(binary_int))}")
+						elif command_header == 10:
+							print('Fetching status')
+							print('Returning status to client')
+						elif command_header == 9:
+							print('Entering power saving mode')
+						else:
+							print('Command not found')
+
+						print(f"Echoing data: {bin(binary_int)}")
+						connection.sendall(data)
```

Comment:
Left the echoing in, this would probably change to be done in the if statements above.